### PR TITLE
[credentialhelper] Add parser for flag syntax

### PR DESCRIPTION
**Pull request by Yannic**
Progress on #15856

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelper.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public final class CredentialHelper {
   }
 
   @VisibleForTesting
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelper.java:51]: You can remove the `@VisibleForTesting`.

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperProvider.java [hunk 2]
```diff
@@ -129,6 +129,23 @@ private void checkHelper(Path path) throws IOException {
           path.isExecutable(), "Credential helper %s is not executable", path);
     }
 
+    /**
+     * Adds a credential helper to use for all {@link URI}s matching the provided pattern, or as
+     * default credential helper if {@code pattern} is empty.
+     *
+     * <p>See {@link #add(String, Path)} for the syntax of {@code pattern}.
+     */
+    public Builder add(Optional<String> pattern, Path helper) throws IOException {
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperProvider.java:138]: IMO it's much clearer for this method to receive a non-optional `pattern`, and call `add(Path)` directly in the case of a default helper.

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/devtools/build/lib/remote/RemoteModuleTest.java [hunk 3]
```diff
@@ -576,4 +583,225 @@ private static void assertRequestMetadata(
...
+            .setWorkspacePath(workspace)
+            .setClientEnvironment(ImmutableMap.of("PATH", pathValue.getPathString()))
+            .setHelperExecutionTimeout(Duration.ZERO)
+            .build();
+    CommandLinePathFactory commandLinePathFactory =
+        new CommandLinePathFactory(
+            fileSystem,
+            ImmutableMap.<String, Path>builder()
+                .put("workspace", workspace)
+                .build());
```
[the remark this answers is not in the document]
Comment (Yannic) [reply src/test/java/com/google/devtools/build/lib/remote/RemoteModuleTest.java:695]: done

File: src/test/java/com/google/devtools/build/lib/remote/RemoteModuleTest.java [hunk 4]
```diff
@@ -576,4 +583,225 @@ private static void assertRequestMetadata(
...
```
[the remark this answers is not in the document]
Comment (Yannic) [reply src/test/java/com/google/devtools/build/lib/remote/RemoteModuleTest.java:607]: done

File: src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java [hunk 5]
```diff
@@ -1130,4 +1134,53 @@ static Credentials newCredentials(
...
+      CredentialHelperEnvironment environment, CommandLinePathFactory pathFactory, String input)
+      throws IOException {
+    Preconditions.checkNotNull(environment);
+    Preconditions.checkNotNull(pathFactory);
+    Preconditions.checkNotNull(input);
+
+    int pos = input.indexOf('=');
+    if (pos > 0) {
+      String scope = input.substring(0, pos);
+      String path = input.substring(pos + 1);
```
[the remark this answers is not in the document]
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java]: done & done

**Comment by Yannic** [comment]
PTAL

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelper.java (hunk 1 above)
Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelper.java:51]: That makes sense, let's keep it.

**Comment by ckolli5** [comment]
@bazel-io fork 5.3.0

**Pull request closed without merging**